core/arch/arm/kernel/secure_partition.c

```diff
@@ -0,0 +1,772 @@
+}
+
+static bool return_helper(bool panic, uint32_t panic_code,
+			  struct thread_svc_regs *svc_regs)
+{
+	if (!panic) {
+		struct tee_ta_session *sess = NULL;
+		struct sec_part_ctx *spc = NULL;
+		size_t n = 0;
+
+		tee_ta_get_current_session(&sess);
+		spc = to_sec_part_ctx(sess->ctx);
+
+		/* Save the return values from StMM */
+		for (n = 0; n <= 7; n++)
+			spc->regs.x[n] = *(&svc_regs->x0 + n);
+
+		spc->regs.sp = svc_regs->sp_el0;
+		spc->regs.pc = svc_regs->elr;
+		spc->regs.cpsr = svc_regs->spsr;
+	}
+
+	svc_regs->x0 = 0;
+	svc_regs->x1 = panic;
+	svc_regs->x2 = panic_code;
+
+	return false;
+}
+
+static void service_compose_direct_resp(struct thread_svc_regs *regs,
+					uint32_t ret_val)
+{
+	uint16_t src_id = 0;
+	uint16_t dst_id = 0;
+
+	/* extract from request */
+	src_id = (regs->x1 >> 16) & UINT16_MAX;
+	dst_id = regs->x1 & UINT16_MAX;
+
+	/* compose message */
+	regs->x0 = FFA_MSG_SEND_DIRECT_RESP_64;
+	/* swap endpoint ids */
+	regs->x1 = SHIFT_U32(dst_id, 16) | src_id;
+	regs->x2 = FFA_PARAM_MBZ;
+	regs->x3 = ret_val;
+	regs->x4 = 0;
+	regs->x5 = 0;
+	regs->x6 = 0;
+	regs->x7 = 0;
+}
+
+/*
+ * Combined read from secure partition, this will open, read and
+ * close the fh
+ */
+static TEE_Result sec_storage_obj_read(unsigned long storage_id, char *obj_id,
+				       unsigned long obj_id_len, void *data,
+				       unsigned long len, unsigned long offset,
+				       unsigned long flags)
+{
+	const struct tee_file_operations *fops = NULL;
+	TEE_Result res = TEE_ERROR_BAD_STATE;
+	struct tee_ta_session *sess = NULL;
+	struct tee_file_handle *fh = NULL;
+	struct sec_part_ctx *spc = NULL;
+	struct tee_pobj *po = NULL;
+	size_t file_size = 0;
+	size_t read_len = 0;
+
+	fops = tee_svc_storage_file_ops(storage_id);
+	if (!fops)
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	if (obj_id_len > TEE_OBJECT_ID_MAX_LEN)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	res = tee_ta_get_current_session(&sess);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	spc = to_sec_part_ctx(sess->ctx);
+	res = tee_mmu_check_access_rights(&spc->uctx,
+					  TEE_MEMORY_ACCESS_WRITE |
+					  TEE_MEMORY_ACCESS_ANY_OWNER,
+					  (uaddr_t)data, len);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	res = tee_pobj_get(&sess->ctx->uuid, obj_id, obj_id_len, flags,
+			   false, fops, &po);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	res = po->fops->open(po, &file_size, &fh);
+	if (res != TEE_SUCCESS)
+		goto err_release;
+
+	read_len = len;
+	res = po->fops->read(fh, offset, data, &read_len);
+	if (res != TEE_SUCCESS) {
+		if (res == TEE_ERROR_CORRUPT_OBJECT) {
+			EMSG("Object corrupt");
+			po->fops->remove(po);
+		}
+		goto err_close;
+	}
+
+	/* make sure we read the entire requested length */
+	if (len != read_len) {
+		res = TEE_ERROR_CORRUPT_OBJECT;
+		goto err_close;
+	}
+
+err_close:
+	po->fops->close(&fh);
+err_release:
+	tee_pobj_release(po);
+
+	return res;
+}
+
+/*
+ * Combined write from secure partition, this will create/open, write and
+ * close the fh
```

Comment:
ditto.

Reply:
I meant: s/fh/object file/.
By the way: s/write from/write to/

Reply:
The from should stay imho. It's StMM that performs the write, so it's coming from the secure partition RPMB driver

Reply:
You're right: `write from` is ok.